### PR TITLE
test(hip): add public-API semantic contract-test suite

### PR DESCRIPTION
**Pull request by mangupta**
## Tracker Info
JIRA ID: AIRUNTIME-2548

## Summary

Adds a **public-API semantic contract-test suite** for HIP under
`projects/hip-tests/catch/contract/`, plus supporting tooling and a CI gate. The
contract layer pins small, portable, device-only semantic guarantees of the
public HIP API (round-trips, accepted-or-unsupported, invalid-input rejection) —
complementing the existing unit/perf/stress tiers rather than duplicating them.

- **591 contract test cases across 118 domains**, covering **483/494** declared
  `hip_runtime_api.h` APIs (~97.8% name-level), each Catch2 `HIP_TEST_CASE` under
  its own domain executable and registered in `contract.yaml`.
- **Coverage drift checker** (`catch/contract/tools/check_contract_coverage.py`)
  + an intentionally-uncovered **allowlist** (`uncovered_apis.txt`, 11 entries
    with reasons), wired into a **PR CI gate**
    (`.github/workflows/hip-contract-coverage.yml`) that fails when a newly-added
    HIP API ships without a contract test or an allowlist entry. Pure static
    analysis — no ROCm/GPU/build required.
- **Generated test-plan inventory** (`catch/TEST_PLAN.md` +
  `catch/tools/gen_test_plan.py`): every case carries an `// @asserts:` intent
  tag; the plan is regenerated from source and staleness-checked in the same
  workflow, to support cross-tier de-duplication as `catch/*` is organized into
  tiers (contract → unit → integration → system → performance → stress).
- **Authoring guide** (`catch/contract/AUTHORING.md` + `TEMPLATE.cc.txt`) with the
  suite conventions (ContractCleanup, image-gating, `HT_AMD`/`BACKEND-DIFF`
  markers, sticky-error clearing, probe-first) so new tests are correct by
  construction.

## Backend portability

The suite is portable across the AMD (ROCm/HIP) backend and the NVIDIA
HIP-over-CUDA backend (`HIP_PLATFORM=nvidia`). AMD-only or CUDA-removed entry
points are `#if HT_AMD`-gated so they compile to empty binaries on NVIDIA rather
than failing to build; real behavioral divergences are marked `BACKEND-DIFF:`.

## Test plan

Validated build + `ctest -L contract` across multiple targets.

## Notes for reviewers

- **PR size**: this branch introduces the full suite in one place, so it exceeds
  the 50-file / 2000-change size guideline. Opened as a **draft** for initial
  review; happy to discuss splitting if preferred.
- **Tracking reference**: no JIRA/issue exists yet — to be added before this
  leaves draft (the PR-bot Title/Description check will flag this until then).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by rakesroy — APPROVED** [review]

File: projects/hip-tests/catch/contract/kernel_launch/test_hip_kernel_launch_contract.cc [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+  REQUIRE(ReadDeviceInt(symbol_ptr) == kExpectedValue);
+}
+
+// @asserts: hipGetSymbolSize - the reported size of a device global array matches its declared byte size
+HIP_TEST_CASE(Contract_KernelLaunch_HipGetSymbolSize_Default_MatchesDeclaredSize) {
+  // Launch a kernel that references the device global array so the symbol is
+  // emitted and resolvable in the device image on every runtime path.
+  hipLaunchKernelGGL(TouchSymbolArrayKernel, dim3(1), dim3(1), 0, 0);
+  HIP_CHECK(hipGetLastError());
```
Comment (g-h-c) [inline projects/hip-tests/catch/contract/kernel_launch/test_hip_kernel_launch_contract.cc:155]: `hipDeviceSynchronize()` should go first then `hipGetLastError()`.

See https://forums.developer.nvidia.com/t/cudadevicesynchronize-not-returning-error-of-type-invalid-configuration-argument/32524

Comment (mangupta) [reply projects/hip-tests/catch/contract/kernel_launch/test_hip_kernel_launch_contract.cc:155]: Updated the comment to reflect the reasoning for the order or GetLastError and DeviceSynchronize.

File: .github/workflows/hip-contract-coverage.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+on:
+  pull_request:
+    types: [opened, synchronize, reopened]
+    paths:
+      - "projects/hip/include/hip/**"
+      - "projects/hip-tests/catch/contract/**"
+      - "projects/hip-tests/catch/config/configs/contract.yaml"
+      - "projects/hip-tests/catch/tools/**"
+      - "projects/hip-tests/catch/TEST_PLAN.md"
+
```
Comment (marbre) [inline .github/workflows/hip-contract-coverage.yml:28]:
```suggestion

permissions:
  contents: read

```

File: .github/workflows/hip-contract-coverage.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+jobs:
+  contract-coverage:
+    runs-on: ubuntu-latest
```
Comment (marbre) [inline .github/workflows/hip-contract-coverage.yml]:
```suggestion
    runs-on: ubuntu-24.04
```
Pin to a specific version as the latest will be bumped at some time we're not in control off.

File: .github/workflows/hip-contract-coverage.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
```
Comment (marbre) [inline .github/workflows/hip-contract-coverage.yml]:
```suggestion
        uses: actions/checkout@3d3c42e5aac5ba805825da76410c181273ba90b1 # v7.0.1
```
`v4` is a floating tag + really outdated. Actions need to be pinned, see https://github.com/ossf/scorecard/blob/main/docs/checks.md#pinned-dependencies for the rational.

File: .github/workflows/hip-contract-coverage.yml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+      - name: Set up Python
+        uses: actions/setup-python@v5
```
Comment (marbre) [inline .github/workflows/hip-contract-coverage.yml]:
```suggestion
        uses: actions/setup-python@5fda3b95a4ea91299a34e894583c3862153e4b97 # v7.0.0
```

As above.

File: .github/workflows/hip-contract-coverage.yml
Comment (marbre) [inline .github/workflows/hip-contract-coverage.yml:1]: Please add the copyright and license header to workflow files as well.

File: projects/hip-tests/catch/contract/array3d/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
+#
+# SPDX-License-Identifier: MIT
```
Comment (marbre) [inline projects/hip-tests/catch/contract/array3d/CMakeLists.txt]: This header seems to be outdated:

```suggestion
# Copyright Advanced Micro Devices, Inc.
# SPDX-License-Identifier: MIT
```

File: projects/hip-tests/catch/tools/gen_test_plan.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
+#!/usr/bin/env python3
+# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
+#
+# SPDX-License-Identifier: MIT
```
Comment (marbre) [inline projects/hip-tests/catch/tools/gen_test_plan.py]:
```suggestion
# Copyright Advanced Micro Devices, Inc.
# SPDX-License-Identifier: MIT
```

File: projects/hip-tests/catch/tools/gen_test_plan.py
Comment (marbre) [inline projects/hip-tests/catch/tools/gen_test_plan.py:1]: Worth adding a test for the tool itself?

**Review by marbre — COMMENTED** [review]
I mainly looked at the workflow changes, the other findings are rather a drive-by and need someone else's review.

File: projects/hip-tests/catch/tools/gen_test_plan.py
Comment (mangupta) [reply projects/hip-tests/catch/tools/gen_test_plan.py:1]: Addressed all the other review comments. If it is alright, I will raise a follow on PR for a test for the tool. I don't want to have to rebase this large PR repeatedly.

File: projects/hip-tests/TESTING.md [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+All tests are C++ built on **Catch2 v3.8.1** and discovered/run through **CTest**.
+Test executables are grouped by CMake `BUILD_*` options and CTest **labels**.
+
+---
+
+# Part A — Current status
+
+## A.1 Test categories (today)
+
+The suite lives under `projects/hip-tests/catch/` and is organized into these
```
Comment (ScottTodd) [inline projects/hip-tests/TESTING.md:50]: nit: want to turn some of these code links into hyperlinks so humans can click them to browse around the source tree?

``[`projects/hip-tests/catch/`](/projects/hip-tests/catch/)`` should create a link that is clickable in both IDEs and on GitHub.

File: .github/workflows/hip-contract-coverage.yml [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+on:
+  pull_request:
+    types: [opened, synchronize, reopened]
+    paths:
+      - "projects/hip/include/hip/**"
+      - "projects/hip-tests/catch/contract/**"
+      - "projects/hip-tests/catch/config/configs/contract.yaml"
+      - "projects/hip-tests/catch/tools/**"
+      - "projects/hip-tests/catch/TEST_PLAN.md"
```
Comment (ScottTodd) [inline .github/workflows/hip-contract-coverage.yml:27]: (For later)

If you want this workflow to be part of a "required check" that blocks merging, you'll need to drop these `paths:` filters and do any filtering in the workflow itself, e.g. with an `if:` condition like here: https://github.com/ROCm/rocm-systems/blob/c0f0844c61ee826377c37c379ed8ab15790b9a83/.github/workflows/therock-ci.yml#L152-L158

As for why, see https://github.com/orgs/community/discussions/44490 - if a check is skipped due to path filters, github still treats it as required.

The "component CI" in rocm-libraries is a good example to follow. It lets each subproject in the repository list the paths it wants to run its own CI on then includes that workflow:
* https://github.com/ROCm/rocm-libraries/blob/develop/.github/workflows/component-ci.yml
* https://github.com/ROCm/rocm-libraries/blob/develop/.github/scripts/component_ci.py
* https://github.com/ROCm/rocm-libraries/blob/develop/.github/workflows/component-ci-miopen.yml

File: projects/hip-tests/catch/contract/tools/check_contract_coverage.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
+CONTRACT_DIR = os.path.dirname(_THIS_DIR)                       # catch/contract
+_CATCH_DIR = os.path.dirname(CONTRACT_DIR)                      # catch
+_HIP_TESTS_DIR = os.path.dirname(_CATCH_DIR)                    # projects/hip-tests
+_PROJECTS_DIR = os.path.dirname(_HIP_TESTS_DIR)                 # projects
+REPO_ROOT = os.path.dirname(_PROJECTS_DIR)                      # <repo>
+
+HEADER_PATH = os.path.join(
+    REPO_ROOT, "projects", "hip", "include", "hip", "hip_runtime_api.h")
+ALLOWLIST_PATH = os.path.join(CONTRACT_DIR, "uncovered_apis.txt")
```
Comment (ScottTodd) [inline projects/hip-tests/catch/contract/tools/check_contract_coverage.py:43]: nit: while the behavior here is good, you may want to switch to using `pathlib` for file system paths instead of the `os.path` module. Pathlib has more natural (and safe) ways to manipulate paths, like
```diff
-HEADER_PATH = os.path.join(REPO_ROOT, "projects", "hip", ...)
+HEADER_PATH = REPO_ROOT / "projects" / "hip" ...
```

* https://docs.python.org/3/library/pathlib.html
* https://github.com/ROCm/TheRock/blob/main/docs/development/style_guides/python_style_guide.md#use-pathlib-for-filesystem-paths

File: projects/hip-tests/catch/contract/tools/check_contract_coverage.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+                        help="path to hip_runtime_api.h (default: repo-relative)")
+    parser.add_argument("--contract-dir", default=CONTRACT_DIR,
+                        help="path to catch/contract (default: repo-relative)")
+    parser.add_argument("--allowlist", default=ALLOWLIST_PATH,
+                        help="path to uncovered_apis.txt (default: repo-relative)")
+    args = parser.parse_args(argv)
+
+    if not os.path.exists(args.header):
+        sys.stderr.write("error: header not found: {}\n".format(args.header))
+        return 2
```
Comment (ScottTodd) [inline projects/hip-tests/catch/contract/tools/check_contract_coverage.py:211]: nit: could raise `FileNotFoundError` here instead of an opaque `return 2`

* https://docs.python.org/3/library/exceptions.html#FileNotFoundError
* https://github.com/ROCm/TheRock/blob/main/docs/development/style_guides/python_style_guide.md#distinguish-between-different-error-conditions

Alternately you could use `parser.error()` as the file not existing is an error in the provided arguments

* https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.error
* https://github.com/ROCm/TheRock/blob/main/docs/development/style_guides/python_style_guide.md#use-argparse-for-cli-flags

File: projects/hip-tests/TESTING.md [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+Tests are labeled from their Catch2 tags (`ADD_TAGS_AS_LABELS`), so `ctest -L
+<label>` selects a feature area or tier when that tag exists (for example,
+`memory`, `stream`, or `contract`; labels are generated from YAML/Catch2 tags,
+not from directory names alone). Each `HIP_TEST_CASE`/`TEST_CASE` becomes one
+CTest entry (one process per case under `catch_discover_tests`). When running
+from the generated `catch_tests/` directory, the build copies `DartConfiguration.tcl`
+there so CTest settings such as the default timeout still apply.
+
+> **Build note:** Catch2 test executables are created `EXCLUDE_FROM_ALL` and are
+> pulled into the build only through their aggregate custom target. The unit
```
Comment (ScottTodd) [inline projects/hip-tests/TESTING.md:174]: FYI with GitHub flavored markdown you can apply some styling to notes like this: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

```
> [!NOTE]
> Useful information that users should know, even when skimming content.
```

> [!NOTE]
> Useful information that users should know, even when skimming content.

File: projects/hip-tests/TESTING.md [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+level. The YAML documents an intended cadence mapping (level 0 for PR/emulation,
+level 1 for nightly), but **CI does not currently set a global level per cadence**
+— wiring that cadence dial to CI is a Part B item (see
+[B.2](#b2-planned-tightening-integration--performance--system)).
+
+## A.4 How to write tests (today)
+
+- **Location & naming:** put a test under `catch/unit/<feature-area>/`; name the
+  file so it is recognized as a test (`test_*` or `*_test.*` per the PR-bot rule),
+  and name each case `<Area>_<Behavior>` so intent is legible in CTest output.
```
Comment (ScottTodd) [inline projects/hip-tests/TESTING.md:205]: You could also enforce some naming scheme via pre-commit or some other tooling that is compatible with local development workflows, rather than just the PR bot that only runs after a pull request is posted. See https://github.com/ROCm/TheRock/blob/4d7626337cac97ed87801f3f57224e12fe509860/.pre-commit-config.yaml#L60-L64 for example

File: projects/hip-tests/TESTING.md [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+  cannot corrupt subsequent cases in the same executable.
+- **Sufficient coverage:** a change to a public API needs at least one test
+  covering the new/changed behavior, including at least one invalid-input path
+  where the API defines one. Register the test's build target and (where used) its
+  YAML config entry so it actually compiles and runs.
+- **References:** `projects/hip-tests/CONTRIBUTING.md` is the repo-wide
+  contribution guide; `projects/hip-tests/catch/README.md` documents the current
+  Catch2 harness mechanics (YAML tags, levels, macros). This file is the testing
+  strategy / tiering overview and should be updated when those lower-level docs
+  change behavior that affects test policy.
```
Comment (ScottTodd) [inline projects/hip-tests/TESTING.md:246]: nit: these could also be turned into hyperlinks to make it easier to follow them

File: projects/hip-tests/TESTING.md [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+- **Structure:** one `TEST_CASE` per behavior; use `SECTION`s for variants. Use
+  `HIP_CHECK` for API calls that must succeed and assert the observable invariant
+  with `REQUIRE`. Clean up every resource you allocate.
+- **Level-aware workload trimming:** use `isQuickLevel()` (from
+  `hip_test_common.hh`) to reduce buffer sizes, iteration counts, or generated
+  parameter sets at `level_0` rather than skipping code paths entirely. Call
+  `isQuickLevel()` inside the test function, not in global/static initializers,
+  because the active level is detected at runtime. If the test uses randomized
+  inputs, use deterministic/reproducible seeds and log or allow overriding the
+  seed so failures can be replayed.
```
Comment (ScottTodd) [inline projects/hip-tests/TESTING.md:215]: Could add a code example/reference for randomized inputs with reproducible seeds here

File: projects/hip-tests/CONTRACT_COVERAGE.md [hunk 16]
```diff
@@ -0,0 +1,784 @@
...
+
+This document tracks the public HIP API semantic contract coverage in `projects/hip-tests/catch/contract`.
+
+The percentages below are approximate API-name coverage against declarations parsed from `projects/hip/include/hip/hip_runtime_api.h`. They are not behavioral coverage: one API can have many modes and edge cases, and the contract layer intentionally covers only small, portable semantic guarantees.
+
+## Snapshot
+
+- Contract tests: 591
+- Declared HIP runtime APIs parsed from `hip_runtime_api.h`: 494
+- Declared HIP runtime APIs directly exercised by contract tests: 486
```
Comment (ScottTodd) [inline projects/hip-tests/CONTRACT_COVERAGE.md:11]: Can you add a date and/or commit hash to this snapshot? The specific numbers here will go stale quickly and listing that detail directly will save people from needing to look at the git history

File: projects/hip-tests/CONTRACT_COVERAGE.md [hunk 17]
```diff
@@ -0,0 +1,784 @@
...
...s accepted-or-unsupported contracts (`hipMemDiscardBatchAsync`, `hipMemDiscardAndPrefetchBatchAsync`, `hipDrvMemDiscardBatchAsync`, `hipDrvMemDiscardAndPrefetchBatchAsync`), each requiring XNACK plus managed or system-allocated memory and reporting `hipErrorNotSupported` where that path is unavailable, plus a null-range rejection. The batch prefetch operation is now covered too (`hipMemPrefetchBatchAsync` accepted-or-unsupported), gated to discrete GPUs because it rejects well-formed inputs with an invalid-value error on an integrated local runtime while the single-range `hipMemPrefetchAsync_v2` and the combined `hipMemDiscardAndPrefetchBatchAsync` accept the identical location; on discrete datacenter GPUs it is accepted. The batch 3D copy is now covered too (`hipMemcpy3DBatchAsync` via a pointer-operand host-device-host round-trip with a byte-scaled single-row/layer extent, accepted-or-unsupported, plus zero-op, null-op-list, and non-zero-flags rejections); the `failIdx` out-parameter is not asserted on because the NVIDIA backend drops it on CUDA 13+. Remaining memory gaps are the driver-style/2D symbol-copy families and array-operand 3D-batch copies not exercised in this harness.
```
Comment (ScottTodd) [inline projects/hip-tests/CONTRACT_COVERAGE.md:735]: This section is a very dense wall of text:

Image

Want to split into sublists or add some paragraph breaks?

**Review by ScottTodd — APPROVED** [review]
I reviewed the .md, .yml, and .py files. I did not review the `_contract.cc` files though I did scan a few of them. No concerns from me with respect to this usage of github actions or the changes to development and testing policies. Most of my comments are just tips for making docs and workflows a bit more helpful.

I think this PR could land as a single change (ideally with squash and merge, as there are 200+ commits), or it could be split into
1. Scaffolding (new docs, yml files, structure for adding the tests)
2. Individual test cases in a single batch or split into a few subtrees)

File: projects/hip-tests/TESTING.md (hunk 8 above)
Comment (mangupta) [reply projects/hip-tests/TESTING.md:50]: good suggestion. will take care of this in the follow on PR

File: .github/workflows/hip-contract-coverage.yml [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+permissions:
+  contents: read
+
+jobs:
+  contract-coverage:
+    runs-on: ubuntu-24.04
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@3d3c42e5aac5ba805825da76410c181273ba90b1 # v7.0.1
```
Comment (geomin12) [inline .github/workflows/hip-contract-coverage.yml:37]: oh rocm-systems is a fairly large repo. we can do sparse-checkout of projects/hip-tests to save some resources / time

**Review by geomin12 — APPROVED** [review]
thanks! just one comment

File: .github/workflows/hip-contract-coverage.yml (hunk 18 above)
Comment (ScottTodd) [reply .github/workflows/hip-contract-coverage.yml:37]: Checkout took 10 seconds here: https://github.com/ROCm/rocm-systems/actions/runs/30813836599/job/91686662146?pr=8827 (the whole job took 14 seconds). That seems fast enough to keep the workflow simple.

**Review by chrispaquot — APPROVED** [review]

File: .github/workflows/hip-contract-coverage.yml (hunk 9 above)
Comment (mangupta) [reply .github/workflows/hip-contract-coverage.yml:27]: yes. we definitely want to make it a required check. @litvaOo : Can you take over fixing the github action workflow

File: projects/hip-tests/catch/contract/tools/check_contract_coverage.py (hunk 10 above)
Comment (mangupta) [reply projects/hip-tests/catch/contract/tools/check_contract_coverage.py:43]: Looks fine. Will take care of this as well.

File: projects/hip-tests/catch/contract/tools/check_contract_coverage.py (hunk 11 above)
Comment (mangupta) [reply projects/hip-tests/catch/contract/tools/check_contract_coverage.py:211]: Agreed

File: projects/hip-tests/TESTING.md (hunk 13 above)
Comment (mangupta) [reply projects/hip-tests/TESTING.md:205]: Will update it in the follow on PR

File: projects/hip-tests/TESTING.md (hunk 14 above)
Comment (mangupta) [reply projects/hip-tests/TESTING.md:246]: Ditto

File: projects/hip-tests/CONTRACT_COVERAGE.md (hunk 16 above)
Comment (mangupta) [reply projects/hip-tests/CONTRACT_COVERAGE.md:11]: Yes. That will be useful. Thanks for the suggestion. Will incorporate it.

File: projects/hip-tests/CONTRACT_COVERAGE.md (hunk 17 above)
Comment (mangupta) [reply projects/hip-tests/CONTRACT_COVERAGE.md:735]: Hmm. Will see why this was generated this way. Definitely should be split into smaller paragraphs.

File: .github/workflows/hip-contract-coverage.yml (hunk 18 above)
Comment (mangupta) [reply .github/workflows/hip-contract-coverage.yml:37]: Will leave it like it is for now, we can monitor for a week and take a call to address this in a follow on PR if needed.

File: projects/hip-tests/TESTING.md (hunk 15 above)
Comment (mangupta) [reply projects/hip-tests/TESTING.md:215]: Sure. Will do that.

File: projects/hip-tests/TESTING.md (hunk 12 above)
Comment (mangupta) [reply projects/hip-tests/TESTING.md:174]: Good suggestion. Will incorporate it into the document.

File: .github/workflows/hip-contract-coverage.yml
Comment (mangupta) [reply .github/workflows/hip-contract-coverage.yml:1]: done.
